### ISSUE TITLE
YoloV4 ttnn.conv2d fails with "Input tensor must be width sharded" error

### ISSUE THREAD

**Issue by keerthana-r-mcw** [tenstorrent/tt-metal#12387]
**Describe the bug**
On implementing the concat op using sharded data as input as in [10808](https://github.com/tenstorrent/tt-metal/issues/10808) [12182](https://github.com/tenstorrent/tt-metal/pull/12182), facing "Input tensor must be width sharded" error in the following conv.

**To Reproduce**
1. Checkout [keerthanar/yolov4_shard_conv_issue](https://github.com/tenstorrent/tt-metal/tree/keerthanar/yolov4_shard_conv_issue)
2. Run `pytest tt-metal/tests/ttnn/unit_tests/operations/test_concat_conv.py::test_concat_conv_yolov4`
3. The following error will occur,
`RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/operations/data_movement/tilize_with_val_padding/device/tilize_with_val_padding_op.cpp:37: input_tensor_a.memory_config().memory_layout == TensorMemoryLayout::WIDTH_SHARDED
E       info:
E       Input tensor must be width sharded`

**Please complete the following environment information:**
- WH n150

**Comment by pavlejosipovic** [comment]
@keerthana-r-mcw  There are two issues with this UT.

1. Input tensor sharding is not setup correctly, this would should up as assert in Debug build
In order to resolve it change
```
output_sharded_memory_config = ttnn.create_sharded_memory_config(
        [512, 128],
        core_grid=ttnn.CoreRangeSet({ttnn.CoreRange((0, 0), (7, 0)), ttnn.CoreRange((0, 1), (4, 1))}),
        strategy=ttnn.ShardStrategy.HEIGHT,
        use_height_and_width_as_shard_shape=True,
    )
```
to 
```
output_sharded_memory_config = ttnn.create_sharded_memory_config(
        [32, 512],
        core_grid=ttnn.CoreRangeSet({ttnn.CoreRange((0, 0), (7, 0)), ttnn.CoreRange((0, 1), (4, 1))}),
        strategy=ttnn.ShardStrategy.HEIGHT,
        use_height_and_width_as_shard_shape=True,
    )
```
2. Second problem is that ttnn::to_layout which is being triggered from with-in conv2d doesn't properly support tilizing height-sharded tensors which need to padded to read multiple of 32 dims. 
This is being addressed in this PR https://github.com/tenstorrent/tt-metal/pull/12641


With these two changes you should be able to run this test-case.

**Comment by keerthana-r-mcw** [comment]
Hi @pavlejosipovic , 
I tried including these two changes, and the unit test is working. But, I am getting low pcc. Can you check?
Thanks.

**Comment by pavlejosipovic** [comment]
@keerthana-r-mcw  seems like you changed test a bit more than what I have suggested can you restore "original" branch by any chance?

**Comment by pavlejosipovic** [comment]
Ok I've tweaked UT a bit to make it work 
Here is my brand with working UT 
 https://github.com/tenstorrent/tt-metal/tree/pjosipovic/yolo4

**Comment by keerthana-r-mcw** [comment]
I'll check and let you know. Thanks!

**Comment by keerthana-r-mcw** [comment]
Hi @pavlejosipovic ,
I checked your branch, and to be clear, when I try to run concat with shard memory config, the conv, which is next to the shard concat is throwing the mentioned error. But the fix you gave doesn't have shard concat. Can you please check once?
Thanks.

**Comment by pavlejosipovic** [comment]
So your use-case is such that you need to run concat on two shared tensors?
If that is the case we should make a UT which tests just that and file a concat bug.
I don't think this is a con related bug.

**Comment by keerthana-r-mcw** [comment]
But the error occurs only when the conv is next to the shard concat.

**Comment by pavlejosipovic** [comment]
Ok I'll try to isolate the concat problem.

**Comment by keerthana-r-mcw** [comment]
Ok Thanks.

**Comment by pavlejosipovic** [comment]
I've pushed new commit to my branch. 
Workaround compared to your original report is now smaller.

UT for concat is added to show-case that problem originates from concat there is a passing and failing version 
```
@pytest.mark.parametrize("should_fail", [False, True])
def test_concat(device, should_fail):
    torch_input = torch.randn(1, 1, 400, 256, dtype=torch.bfloat16).float()
    goldern_concat = torch.cat((torch_input, torch_input), dim=3)

    ttnn_input = ttnn.from_torch(torch_input, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)
    output_sharded_memory_config = ttnn.create_sharded_memory_config(
        [32, 512],
        core_grid=ttnn.CoreRangeSet({ttnn.CoreRange((0, 0), (7, 0)), ttnn.CoreRange((0, 1), (4, 1))}),
        strategy=ttnn.ShardStrategy.HEIGHT,
        use_height_and_width_as_shard_shape=True,
    )

    if should_fail:
        ttnn_concat_result = ttnn.concat([ttnn_input, ttnn_input], dim=3, memory_config=output_sharded_memory_config)
    else:
        ttnn_concat_result = ttnn.concat([ttnn_input, ttnn_input], dim=3)
        ttnn_concat_result = ttnn.to_memory_config(ttnn_concat_result, output_sharded_memory_config)
    ttnn_output_as_torch = ttnn.to_torch(ttnn_concat_result)

    pcc = 0.99
    passing, pcc_msg = check_with_pcc(goldern_concat, ttnn_output_as_torch, pcc=pcc)
    logger.info(f"PCC = {pcc_msg}. Threshold = {pcc}")
    assert passing
```

**Comment by keerthana-r-mcw** [comment]
Hi @pavlejosipovic ,
This UT reproduces the low PCC issue. 
Will you be working on this issue further or should this be reassigned to someone else?
CC: @dvartaniansTT

**Comment by pavlejosipovic** [comment]
I'll merge the current PR I have for this, and for this concat issue we should open a new one and assign it to the concat op owner.

**Comment by pavlejosipovic** [comment]
But I think that you are not blocked by this so new issue shouldn't be P0, right?

**Comment by keerthana-r-mcw** [comment]
We already have a working yolov4 pipeline with interleaved concat, but we are optimising the pipeline which is currently given high priority. Therefore, since we want this shard concat for optimisation, the issue should be P0.

**Comment by keerthana-r-mcw** [comment]
FYR @punithsekar

**Comment by punithsekar** [comment]
FYR @dvartaniansTT

**Comment by mywoodstock** [comment]
Please create a P0 issue for concat and assign to @tarafdarTT . We can then close this one.

**Comment by tarafdarTT** [comment]
> Please create a P0 issue for concat and assign to @tarafdarTT . We can then close this one.

@keerthana-r-mcw , if the bug is concat specific please make a concat only unit test and link the branch in it. I looked at the test in the above branch, there is a lot going on there.

**Comment by tarafdarTT** [comment]
NVM @keerthana-r-mcw I see the concat there.

**Comment by mywoodstock** [comment]
@keerthana-r-mcw has the new ticket for concat been created? we can close this one then.

**Comment by keerthana-r-mcw** [comment]
Closing this issue since new ticket is created https://github.com/tenstorrent/tt-metal/issues/13037
